### DISCUSSION TITLE
C++ Compatibility Suggestions for easy migration from C++ to C#

### DISCUSSION THREAD

**Discussion by SabaRG** [dotnet/csharplang#6983] [Language Ideas]
I want to implement the iostream class of C++ in C#, especially the behaviours of the cin and the cout objects. I have two problems:
1- I can easily simulate the cout behaviour using the following code:

```
// Implementation of ofstream or ostream in C++ using C#
public class OStream {

        TextWriter tw;

        public override Encoding Encoding => throw new NotImplementedException();

        public OStream(TextWriter tw)
        {
            this.tw = tw;
        }

        public static OStream operator <<(OStream left, bool right)
        {
            left.tw.Write(right);
            return left;
        }

        public static OStream operator <<(OStream left, int right)
        {
            left.tw.Write(right);
            return left;
        }

        public static OStream operator <<(OStream left, double right)
        {
            left.tw.Write(right);
            return left;
        }

        public static OStream operator <<(OStream left, string right)
        {
            left.tw.Write(right);
            return left;
        }

        public void Close()
        {
            tw.Close();
        }
}
````

but there is a problem while using the cout as the following:
```
var outs = new OStream(Console.Out);
outs << 1 << "salam" << true;
```
the error is:

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	CS0201	Only assignment, call, increment, decrement, await, and new object expressions can be used as a statement	Program.cs	18	Active

**I'm asking to disable CS0201 error for operator << and >>.** Even though I know that I can use this silly syntax to avoid CS0201 error:
`_ = outs << 1 << "salam" << true;`

2- when I want to implement the cin behaviour as the following code: 
```
    public class IStream {

        TextReader tr;

        public IStream(TextReader tr)
        {
            this.tr = tr;
        }

        public static IStream operator >>(IStream left, out int right)
        {
            right = 0;
            while (true)
            {
                try
                {
                    int c = left.tr.Read();
                    if (char.IsDigit((char)c))
                    {
                        right = 10 * right + Convert.ToInt32(c - '0');
                    }
                    else
                    {
                        break;
                    }
                }
                catch (Exception)
                {
                    break;
                }
            }
            return left;
        }
    }
```
I will receive the following error:
> Severity	Code	Description	Project	File	Line	Suppression State
> Error	CS0631	ref and out are not valid in this context	DeterministicAnnealing_DARP	iostream.cs	19	Active

**Is it possible to let using out and ref variables in operator overloading without needing to strictly use those keywords in calling time for operator overloading?**

for the record, I implement istream (or ifstream) as the following:

```
    public class IStream
    {
        TextReader tr;

        public IStream(TextReader tr)
        {
            this.tr = tr;
        }

        public bool? EndOfStream
        {
            get
            {
                if (tr is StreamReader)
                    return (tr as StreamReader).EndOfStream;
                else
                    return null;
            }
        }

        public IStream SkipSpace()
        {
            while (EndOfStream == false)
            {

                var c = tr.Peek();
                if (c < 0)
                {
                    break;
                }
                if (char.IsWhiteSpace((char)c))
                    tr.Read();
                else
                    break;
            }
            return this;
        }

        public IStream Next(out int right)
        {
            right = 0;
            SkipSpace();
            var sb = new StringBuilder();
            while (true)
            {
                int c = tr.Peek();
                try
                {
                    char c2 = (char)c;
                    if (c < 0 || char.IsWhiteSpace(c2))
                    {
                        break;
                    }
                    else if (char.IsDigit(c2) || c2 == '-' || c2 == '+')
                    {
                        sb.Append(c2);
                        tr.Read();
                    }
                }
                catch (Exception)
                {
                    break;
                }
            }
            right = Convert.ToInt32(sb.ToString());
            return this;
        }

        public IStream Next(out double right)
        {
            right = 0;
            SkipSpace();
            var sb = new StringBuilder();
            while (true)
            {
                int c = tr.Peek();
                try
                {
                    char c2 = (char)c;
                    if (c < 0 || char.IsWhiteSpace(c2))
                    {
                        break;
                    }
                    else if (char.IsDigit(c2) || c2 == '.' || c2 == '-' || c2 == '+')
                    {
                        sb.Append(c2);
                        tr.Read();
                    }
                }
                catch (Exception)
                {
                    break;
                }
            }
            right = Convert.ToDouble(sb.ToString());
            return this;
        }

        public IStream Next(out string right)
        {
            right = "";
            SkipSpace();
            var sb = new StringBuilder();
            while (true)
            {
                int c = tr.Peek();
                try
                {
                    char c2 = (char)c;
                    if (c < 0 || char.IsWhiteSpace(c2))
                    {
                        break;
                    }
                    else
                    {
                        sb.Append(c2);
                        tr.Read();
                    }
                }
                catch (Exception)
                {
                    break;
                }
            }
            right = sb.ToString();
            return this;
        }

        //public static IStream operator >>(IStream left, out int right)
        //{
        //    left.Next(out right);
        //    return left;
        //}
    }
```

and it can be used like this:

```
            var cin = new IStream(Console.In);
            int x;
            double y;
            string z;
            cin.Next(out x).Next(out y).Next(out z);
```
Even though I prefer to use it like this:
`cin >> x >> y >> z;`

#iostream #istream #ostream #ifstream #ofstream #fstream #C++ #C#

**Comment by HaloFour** [comment]
C#'s limitations on operator overloading were done explicitly to limit what was considered "abuse" of said operators in C++.  I don't know if streams were considered on the list of abusers, but I'm not sure why there would be value in trying to implement iostream in C#.  The BCL has it's own APIs for interacting with the console and with streams.

**Reply by SabaRG** [reply]
As I said in the title, it is necessary to easily convert (and migrate) a C++ code into C#.
If you think it is not important for C++ codes that can be easily used in C#, it's ok for me because I know both languages and I can easily reproduce a C++ code using C#, but many coders can not do it. Besides, cin and cout are more easy to use for many users instead of Console.Read or Console.Write, especially cin that let multiple variables reading in single or multi-line from console like:
```
int x;
float y;
string z;
cin>>x>>y>>z;
```
> sample input 1: 12 18.4 ali 
> sample input 2: 12\n18.4\nali

It is hard to read the sample 1 in C#. Don't you think?

**Reply by HaloFour** [reply]
Given the number of other differences between C++ and C#, I personally don't think that replicating iostream will move the needle much in helping folks.

The features of `cin`, when it comes to parsing data, aren't dependent on the syntax.  The limitations imposed on those operators are intended to discourage using them for cases other than bit-shifting.

**Reply by SabaRG** [reply]
I can't understand this sentence:

> The features of cin, when it comes to parsing data, aren't dependent on the syntax.

Why should be dependent on the syntax? Why should our syntax be different when we read an integer or string or double? Why we shouldn't can have the same syntax for all of them and use IO simply?

I respect your point, but I can't entirely agree with it. **We** define the meaning of each operator. If C# wants to always use >> just for bit shifting (or + just for addition and so on), it never lets other developers overload operators like java.

**Reply by HaloFour** [reply]
> Why should be dependent on the syntax? Why should our syntax be different when we read an integer or string or double? Why we shouldn't can have the same syntax for all of them and use IO simply?

I mean that you can achieve the same functionality of `int x; float y; string z; cin>>x>>y>>z;` without using bit shifting operators.  It doesn't require different syntax per type.

> I respect your point, but I can't entirely agree with it. We define the meaning of each operator. If C# wants to always use >> just for bit shifting (or + just for addition and so on), it never lets other developers overload operators like java.

Because the team still wants to enable bit shifting for custom numeric types.  The designers of C# explicitly wanted to avoid the mess C++ became through weird applications of operator overloading.  You're free to disagree with it, but I think you'd need a much better argument than claiming to want to do exactly what the language designers don't want to allow you to do.

**Comment by FaustVX** [comment]
You can use a discard to allow that, but like @HaloFour said, it's still an abuse of operators.
`_ = outs << 1 << "salam" << true;`
[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwgBuE2YBTAJiAGoAPgAEATAEYAsACgRABgAEIiQDoAkgHkA3LIXK1AFQgAPGLpmy2+RSTgxkigLyKyEAO6LNAZSrQcABSkFCQANhCqmvYAlBayAPrOtvaOADypihKK6YoARMhQoVA4udkZVHAQcXIAzMpiXr5IxYogisZmAOrYMJyKAN6ysooj7aYw3Vi9CIow7tWjynUkHAjYvBCKAKJkNCS8WGQA5tu7+4cnTgB8swAWCCSebp4AciQw6jgADuE4EGS9XhbEw0CBfGBYcgBWJDGSLER1Hx+YoBDoTHp9ObRYajQZwxajGC3LDIVRzJJzCyLAC+sPhdRUADZGsicLYvpxYAwygEkc02eEwDAmIpgCQwopsEdbjBsfiRniCaNBTAye5VJNegEpTKYfKCSIAOyKFVU0a0yz6hEGZl8/zszn0GbpXlNe0qkWHGCSrDS2U4hUBgkqtUajHa326s1Ko0miBC6OKC1B61Mln8h0ILnO1Ku1lxoUilRKHX+/WKpUF1VzMNTCARv16yvKY2moMW6lAA)

**Reply by HaloFour** [reply]
I didn't realize that relaxation was implemented.  I know it was discussed at one point:

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/relaxing_shift_operator_requirements

But note that allowing developers to implement iostream-like syntax was explicitly mentioned in the drawbacks.

**Reply by FaustVX** [reply]
IIRC, it was done for the generic math feature

**Reply by SabaRG** [reply]
> You can use a discard to allow that, but like @HaloFour said, it's still an abuse of operators. `_ = outs << 1 << "salam" << true;` [sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwgBuE2YBTAJiAGoAPgAEATAEYAsACgRABgAEIiQDoAkgHkA3LIXK1AFQgAPGLpmy2+RSTgxkigLyKyEAO6LNAZSrQcABSkFCQANhCqmvYAlBayAPrOtvaOADypihKK6YoARMhQoVA4udkZVHAQcXIAzMpiXr5IxYogisZmAOrYMJyKAN6ysooj7aYw3Vi9CIow7tWjynUkHAjYvBCKAKJkNCS8WGQA5tu7+4cnTgB8swAWCCSebp4AciQw6jgADuE4EGS9XhbEw0CBfGBYcgBWJDGSLER1Hx+YoBDoTHp9ObRYajQZwxajGC3LDIVRzJJzCyLAC+sPhdRUADZGsicLYvpxYAwygEkc02eEwDAmIpgCQwopsEdbjBsfiRniCaNBTAye5VJNegEpTKYfKCSIAOyKFVU0a0yz6hEGZl8/zszn0GbpXlNe0qkWHGCSrDS2U4hUBgkqtUajHa326s1Ko0miBC6OKC1B61Mln8h0ILnO1Ku1lxoUilRKHX+/WKpUF1VzMNTCARv16yvKY2moMW6lAA)

Yes, I know, but it is a foolish syntax. So, I suggest to remove that restriction to avoid this syntax.

**Reply by SabaRG** [reply]
> I didn't realize that relaxation was implemented. I know it was discussed at one point:
> 
> https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/relaxing_shift_operator_requirements
> 
> But note that allowing developers to implement iostream-like syntax was explicitly mentioned in the drawbacks.

Yes, I know, and that's the reason of my suggestion to let implement cin and cout. Before that relaxation, we couldn't implement cout as I wrote, but now we can. Just there is a problem as I mentioned in the main thread.

**Reply by FaustVX** [reply]
To me, it seems okay.
Having `a << b;` or `a >> a;` is like having `a + b;` it looks weird.

**Comment by iam3yal** [comment]
@SabaRG In the OP you're saying you want to implement the **behaviour** of `cin` and `cout` but then you actually care about the syntax, don't you think that using something like this `OStream.Use(Console.Out).Write(1).Write("salam").Write(true);` is more readable than the usage of operators? also it's not clear why you're using both composition and inheritance, you don't have to inherit from `TextWriter`.

**Reply by SabaRG** [reply]
Thanks for your hint, I remove the inheritance now in the main thread because I wrote the code using inheritance the first time and then I have to use composition to allow users to send their stream (file stream or console stream).

Respectly, I prefer `cin>>x>>y>>z;` instead of:
```
var tmp = Console.ReadLine().Split();
x = Convert.ToInt32(tmp[0]);
y = Convert.ToSingle(tmp[1]);
z = tmp[2];
```
Also, I prefer `cout<<x<<y<<z;` instead of `cout.Write(1).Write("salam").Write(true);`. Sorry, It depends on developers' opinion and when C# is going to support many simple syntaxes from other programming languages like python, why can't let developers use C++ syntax?

**Reply by HaloFour** [reply]
C# doesn't support syntax from Python.  It may borrow features from other languages, but it will do so using syntax that fits with C#.  It's not a goal to make C# syntax look like the syntax of any other language.  It did borrow a lot of syntax from C++ in order to achieve a familiar feel, but iostream and arbitrary operator overloading was explicitly disallowed as it was considered confusing and undesirable.

**Reply by FaustVX** [reply]
Personally, I don't know very much C++, but reading `cin >> x >> y >> z;` I don't know what it means, I need to read the code below to understand it.
C# is known to be mainly explicit rather than concise.
So having
```C#
var tmp = Console.ReadLine().Split();
x = int.Parse(tmp[0]);
y = float.Parse(tmp[1]);
z = tmp[2];
```
 is much more idiomatic to C#

**Reply by SabaRG** [reply]
> C# doesn't support syntax from Python. It may borrow features from other languages, but it will do so using syntax that fits with C#. It's not a goal to make C# syntax look like the syntax of any other language. It did borrow a lot of syntax from C++ in order to achieve a familiar feel, but iostream and arbitrary operator overloading was explicitly disallowed as it was considered confusing and undesirable.

I add my implementation for `cin` in the main thread. It's not my decision. If the team like it, can let it, otherwise, I will use my implementation as before.

**Reply by SabaRG** [reply]
> idiomatic
The `cin >> x >> y >> z;` meaning is very simple: "read the console input (c in) and write the values in variables x, y, and z. The data type of variables is illustrative enough that how the input stream should be read and parsed.

As you know, the newer version of C# is not very idiomatic as many people like, e.g. look at the following features:

```
// Default in deconstruction (Not C# 8, Changed to C#Next):

(int x, string y) = (default, default);    // C# 7  
(int x, string y) = default;               // C# 8  

// Null Coalescing Assignment: 
if (variable == null)  
{  
  variable = expression; // C# 1..7  
}  
variable ??= expression; // C# 8 

// Static Local Functions:
int AddFiveAndSeven()  
{  
  int y = 5; int x = 7;  
  return Add(x, y);  
  static int Add(int left, int right) => left + right;  
}  
```

**Reply by SabaRG** [reply]
> C# doesn't support syntax from Python. It may borrow features from other languages, but it will do so using syntax that fits with C#. It's not a goal to make C# syntax look like the syntax of any other language. It did borrow a lot of syntax from C++ in order to achieve a familiar feel, but iostream and arbitrary operator overloading was explicitly disallowed as it was considered confusing and undesirable.

The no-main-method feature in C# is derived from many other programming languages like Python. Besides, Tuples, records, and many other new features come to let C# can be used easily like Python or Javascript (or Typescript).

**Reply by CyrusNajmabadi** [reply]
> and when C# is going to support many simple syntaxes from other programming languages like python, why can't let developers use C++ syntax?

It's not a goal to allow all syntaxes from other languages. We add syntax to the language we can best fits the language and our own goals for it.

**Reply by HaloFour** [reply]
> Besides, Tuples, records, and many other new features come to let C# can be used easily like Python or Javascript (or Typescript).

All of those features were adapted to fit C#.  C# never simply takes syntax from other languages.

Even if the team were to relent here and allow bit-shifting operators in statements that wouldn't address the billion other ways in which C# is completely unlike C++.

**Reply by CyrusNajmabadi** [reply]
We have no desire to intentionally enable scenarios where operators are side effecting. If you want to use it that way, we won't stop you. But we're not going to invest in making that experience more pleasant.

**Reply by HaloFour** [reply]
Importantly, .NET is intended to be a polyglot runtime.  If you want C++ syntax with .NET functionality and C# interoperability, Microsoft already ships a compiler to support that:  [C++/CLI](https://learn.microsoft.com/en-us/cpp/dotnet/dotnet-programming-with-cpp-cli-visual-cpp).

**Reply by SabaRG** [reply]
> We have no desire to intentionally enable scenarios where operators are side effecting. If you want to use it that way, we won't stop you. But we're not going to invest in making that experience more pleasant.

@CyrusNajmabadi 
Sorry, I don't wanna argue, but it seems almost all repliers are not familiar with C# and they just want to discuss and argue with the proposed feature! Did you know that operator overloading in C# can have side effects now? Look at this example to learn:

```
    class Test
    {
        int i;

        public static Test operator +(Test a, Test b)
        {
            a.i = 12;  // side effect on object a
            b.i = 13; // side effect on object b
            return new Test() { i = a.i + b.i };
        }
    }
```
Besides, I can implement `cin` with this idea (using an object for the destination variable), but the user will be bothered to use it like:
```
    class Data<T>
    {
        public T data { get; set; }
    }
...
Data<int> d;
cin>>d;
int x = d.data;
```
Even though as you know, the user has to write: `_ = cin>>t;`

**Reply by FaustVX** [reply]
@SabaRG 
> Did you know that operator overloading in C# can have side effects now?

I'm pretty sure a lot of C# developer know operators can have side-effects. (just like implicit cast)
After all, operators are just regular methods with special syntax.
But it's expected to not have side-effects on operators, just like implicit cast, it's expected to not lose data after (casting an `int` to a `long`, but not in the opposite order)
Given your example, what is the result of `t` after `var t = new Test(); t += t;` ?
Answer here: [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUAG4TIAECtAvLTAKYDutAKqwM4IAKAJQBuCg2zMEY8ggB0UOQEkYvAA6sAxoNEUi3PggoBvCrXO18AZlqwGUGWYvXLxAGwH+tAPYbkEBG86bAEeLwg4TwZgISdzU3ILJNoIBSZaYkIRcwB6HNpeKAxWWlYAMzKtBm8YH2AAKyqUuOTgNOZiK2y8gqKS8srtH1rvBqbgFqT8AHYWDijhWmNbdNSoWmxaNvWAXxkknYodoA)

**Reply by CyrusNajmabadi** [reply]
> Did you know

Yes.  I'm one of the language designers here :)

That doesn't change at all what i said, which is that we have no desire to *intentionally* enable scenarios where they are side-effecting.  We have no doubt people can abuse what is already there.  But that is different from going out of our way to support the abuse.

**Reply by SabaRG** [reply]
> @SabaRG
> 
> > Did you know that operator overloading in C# can have side effects now?
> 
> I'm pretty sure a lot of C# developer know operators can have side-effects. (just like implicit cast) After all, operators are just regular methods with special syntax. But it's expected to not have side-effects on operators, just like implicit cast, it's expected to not lose data after (casting an `int` to a `long`, but not in the opposite order) Given your example, what is the result of `t` after `var t = new Test(); t += t;` ? Answer here: [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUAG4TIAECtAvLTAKYDutAKqwM4IAKAJQBuCg2zMEY8ggB0UOQEkYvAA6sAxoNEUi3PggoBvCrXO18AZlqwGUGWYvXLxAGwH+tAPYbkEBG86bAEeLwg4TwZgISdzU3ILJNoIBSZaYkIRcwB6HNpeKAxWWlYAMzKtBm8YH2AAKyqUuOTgNOZiK2y8gqKS8srtH1rvBqbgFqT8AHYWDijhWmNbdNSoWmxaNvWAXxkknYodoA)

If they know, then why you said the operators should not have side effects?!
Besides, If the PL designers want to disallow side effects, they should let `const` or `readonly` for parameters of operator overloading methods that unfortunately they don't let! I checked it before, just for the record look at the following example:
```
        public static OStream operator <<(OStream left, const Data<T> right)
        {
            left.tw.Write(right.ToString());
            return left;
        }
```
the above code has these errors:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1031	Type expected	TestProject	iostream.cs	177	Active
Error	CS1001	Identifier expected	TestProject	iostream.cs	177	Active
Error	CS1026	) expected	TestProject	iostream.cs	177	Active
Error	CS1002	; expected	TestProject	iostream.cs	177	Active
Error	CS0145	A const field requires a value to be provided	TestProject	iostream.cs	177	Active
Error	CS1003	Syntax error, ',' expected	TestProject	iostream.cs	177	Active
Error	CS1002	; expected	TestProject	iostream.cs	178	Active
Error	CS1519	Invalid token '(' in class, record, struct, or interface member declaration	TestProject	iostream.cs	179	Active
Error	CS8124	Tuple must contain at least two elements.	TestProject	iostream.cs	179	Active
Error	CS1519	Invalid token ';' in class, record, struct, or interface member declaration	TestProject	iostream.cs	179	Active
Error	CS1519	Invalid token ';' in class, record, struct, or interface member declaration	TestProject	iostream.cs	180	Active
Error	CS1519	Invalid token ';' in class, record, struct, or interface member declaration	TestProject	iostream.cs	180	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	TestProject	iostream.cs	188	Active
Error	CS0501	'OStream.operator <<(OStream, ?)' must declare a body because it is not marked abstract, extern, or partial	TestProject	iostream.cs	177	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	TestProject	iostream.cs	183	Active
```
I hope, the PL designers at least let `const` or `readonly` or whatever for operator overloading or all other general methods.

**Reply by FaustVX** [reply]
> If they know, then why you said the operators should not have side effects?!

I don't understand, it's not because something **can** do something than they **should** do it.

**Comment by DavidArno** [comment]
> and it can be used like this:
```cs            
            var cin = new IStream(Console.In);
            int x;
            double y;
            string z;
            cin.Next(out x).Next(out y).Next(out z);
```            
You can write that as:
```cs
var cin = new IStream(Console.In);
cin.Next(out int x).Next(out double y).Next(out string z);
```
It's still looks like C# code written by a newcomer to the language who's trying to force another language's idioms on their new code though.

> Even though I prefer to use it like this:
```cpp
    cin >> x >> y >> z;
```
Then stick with C++. Seriously. When learning a new language, one of the most important things a developer can do is leave all preconceived ideas on syntax and idioms at the door and start afresh., Learn the idioms of the new language as early as possible and use them. It ensures that you start writing code that others can easily read as quickly as possible. And those "others" includes yourself in three years time, when you'll come back to this code and (hopefully) wonder what on earth you were thinking trying to mangle the language in this way.

**Reply by SabaRG** [reply]
> > and it can be used like this:
> 
> ```cs
>             var cin = new IStream(Console.In);
>             int x;
>             double y;
>             string z;
>             cin.Next(out x).Next(out y).Next(out z);
> ```
> 
> You can write that as:
> 
> ```cs
> var cin = new IStream(Console.In);
> cin.Next(out int x).Next(out double y).Next(out string z);
> ```
> 
> It's still looks like C# code written by a newcomer to the language who's trying to force another language's idioms on their new code though.
> 
> > Even though I prefer to use it like this:
> 
> ```c++
>     cin >> x >> y >> z;
> ```
> 
> Then stick with C++. Seriously. When learning a new language, one of the most important things a developer can do is leave all preconceived ideas on syntax and idioms at the door and start afresh., Learn the idioms of the new language as early as possible and use them. It ensures that you start writing code that others can easily read as quickly as possible. And those "others" includes yourself in three years time, when you'll come back to this code and (hopefully) wonder what on earth you were thinking trying to mangle the language in this way.

Thanks for your efforts, but I'm a more than 10 years old C# developer. I see that nowadays the C# syntax is changing a lot, and I want to share some ideas. 
Never mind.
Sincerely

**Reply by DavidArno** [reply]
> I see that nowadays the C# syntax is changing a lot, and I want to share some ideas. Never mind. Sincerely

"*All new features added to C++ are intended to fix previously new features added to C++ – David Jameson*"
"*C++: an octopus made by nailing extra legs onto a dog. – Steve Taylor*"
"*[keeping somebody] from using C++ makes me feel like I saved a life – aiju*"

A common theme I find amongst a great many developers is that the last place that we should look for inspiration for new language ideas is within C++. I appreciate you likely disagree, but I think you are in the minority within the developer community. As HaloFour mentioned at the beginning, the original C# designers set out to protect the language from what many see as the abuse of operators that is common in C++. It is extremely unlikely that the current team would be persuaded to abandon that principle to allow deliberate abuse of operators at this stage therefore.
